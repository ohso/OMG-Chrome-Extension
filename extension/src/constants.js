// SPDX-License-Identifier: GPL-3.0-or-later
// SPDX-FileCopyrightText: Copyright (c) 2024 Sam Tran
export const FEED_URL = '';

export const HOME_URL = '';

export const TITLE = 'OMG! Ubuntu!';

/**
 * 15 minutes in minutes
 * @type {number}
 */
export const POLL_INTERVAL = 15;

export const COMPAT_LAST_NOTIFICATION = 'compatLastNotification';
export const COMPAT_ARTICLES = 'compatArticles';

export const FETCH_ALARM = 'fetch-alarm';

export const INITIALIZED_KEY = 'initialized';
export const ARTICLES_KEY = 'articles';
export const LAST_NOTIFICATION_KEY = 'lastNotification';
