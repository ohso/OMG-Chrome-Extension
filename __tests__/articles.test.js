// SPDX-License-Identifier: GPL-3.0-or-later
// SPDX-FileCopyrightText: Copyright (c) 2024 Sam Tran
import { afterEach, beforeEach, describe, expect, it, jest } from '@jest/globals';

import {
  fetchArticles,
  markAsRead,
  markAllAsRead,
  decodeHtmlEntities,
} from '../extension/src/articles.js';
import { mockArticles, mockResponse } from '../mock.js';
import { ARTICLES_KEY } from '../extension/src/constants.js';

describe('articles', () => {
  beforeEach(() => {
    jest.spyOn(global, 'fetch').mockResolvedValue({
      json: () => mockResponse,
    });
    jest.spyOn(global.chrome.storage.sync, 'get').mockResolvedValue(Promise.resolve({}));
    jest.spyOn(global.chrome.storage.sync, 'set').mockImplementation(jest.fn());
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('fetchArticles', () => {
    it('should fetch and store all articles on first run', async () => {
      await fetchArticles();

      expect(global.chrome.storage.sync.set.mock.calls).toMatchSnapshot();
    });

    it('should merge new articles with existing articles', async () => {
      jest.spyOn(global.chrome.storage.sync, 'get').mockResolvedValue(Promise.resolve({ [ARTICLES_KEY]: mockArticles }));

      await fetchArticles();

      expect(global.chrome.storage.sync.set.mock.calls).toMatchSnapshot();
    });

    it('should handle missing thumbnails', async () => {
      jest.spyOn(global, 'fetch').mockResolvedValue({
        json: () => [{
          ...mockResponse[0],
          _embedded: {},
        }],
      });
      jest.spyOn(global.chrome.storage.sync, 'get').mockResolvedValue(Promise.resolve({}));

      await fetchArticles();

      expect(global.chrome.storage.sync.set.mock.calls).toMatchSnapshot();
    });

    it('should gracefully handle network failures', async () => {
      jest.spyOn(global, 'fetch').mockRejectedValue(new Error('Network error'));

      expect(fetchArticles()).resolves.not.toThrow();
    });
  });

  describe('markAsRead', () => {
    it('should mark a single found article as read', async () => {
      jest.spyOn(global.chrome.storage.sync, 'get').mockResolvedValue(Promise.resolve({
        [ARTICLES_KEY]: mockArticles.map(a => ({
          ...a,
          unread: true,
        })),
      }));

      await markAsRead(mockArticles[1].id);

      expect(global.chrome.storage.sync.set.mock.calls[0][0].articles.map((a) => a.unread)).toMatchInlineSnapshot(`
        [
          true,
          false,
          true,
          true,
          true,
          true,
          true,
          true,
          true,
          true,
          true,
        ]
      `);
    });

    it('should handle no found articles', async () => {
      expect(markAsRead(1234)).resolves.not.toThrow();
    });
  });

  describe('markAllAsRead', () => {
    it('should mark all articles as read', async () => {
      jest.spyOn(global.chrome.storage.sync, 'get').mockResolvedValue(Promise.resolve({
        [ARTICLES_KEY]: mockArticles.map(a => ({
          ...a,
          unread: true,
        })),
      }));

      await markAllAsRead();

      expect(global.chrome.storage.sync.set.mock.calls[0][0].articles.every(a => !a.unread)).toBe(true);
    });
  });

  describe('decodeHtmlEntities', () => {
    it.each([
      ['&amp;', '&'],
      ['&quot;', '"'],
      ['&apos;', '\''],
      ['&lt;', '<'],
      ['&gt;', '>'],
      [
        'IBM&#8217;s Iconic Model-M Keyboard Reimagined by 8BitDo"',
        'IBM\'s Iconic Model-M Keyboard Reimagined by 8BitDo"'],
      [
        'Track Time Differently with &#8216;Day Progress&#8217; for GNOME Shell',
        'Track Time Differently with \'Day Progress\' for GNOME Shell',
      ],
      [
        'Microsoft 365 Account Issues in Ubuntu 24.04? Here&#8217;s a Fix',
        'Microsoft 365 Account Issues in Ubuntu 24.04? Here\'s a Fix',
      ],
      [
        'VMware Workstation Pro Now Free to Use on Linux &amp; Windows',
        'VMware Workstation Pro Now Free to Use on Linux & Windows',
      ],
      [
        'Linux Kernel 6.9 Released — And It&#8217;s Packed with Improvements',
        'Linux Kernel 6.9 Released — And It\'s Packed with Improvements',
      ],
      [
        'Google&#8217;s Flutter Team Layoffs Leave Ubuntu Devs in a Flap',
        'Google\'s Flutter Team Layoffs Leave Ubuntu Devs in a Flap',
      ],
      [
        'Disqus Comments Not Loading? It&#8217;s Not Me Bro, It&#8217;s Firefox',
        'Disqus Comments Not Loading? It\'s Not Me Bro, It\'s Firefox',
      ],
      [
        'Mozilla Devs Adding &#8216;New Tab Wallpapers&#8217; to Firefox',
        'Mozilla Devs Adding \'New Tab Wallpapers\' to Firefox',
      ],
      [
        'VLC Media Player Hits 5 Billion Downloads &#8211; Big Changes Ahead',
        'VLC Media Player Hits 5 Billion Downloads - Big Changes Ahead',
      ],
      [
        'Linux Mint 22: Codename &#038; New Cinnamon Feature Revealed',
        'Linux Mint 22: Codename & New Cinnamon Feature Revealed',
      ],
      [
        'KDE Wallpaper Competition Kicks-Off  &#8211; Win a 13-inch Framework Laptop',
        'KDE Wallpaper Competition Kicks-Off  - Win a 13-inch Framework Laptop',
      ],
      [
        'Fix &#8220;Target Configured Multiple Times&#8221; Error on Ubuntu',
        'Fix "Target Configured Multiple Times" Error on Ubuntu',
      ],
      [
        '&#8216;Burn My Windows&#8217; &#038; &#8216;Desktop Clock&#8217; GNOME Extensions Updated',
        '\'Burn My Windows\' & \'Desktop Clock\' GNOME Extensions Updated',
      ],
    ])('should decode %s correctly', (entity, expected) => {
      expect(decodeHtmlEntities(entity)).toBe(expected);
    });
  });
});
