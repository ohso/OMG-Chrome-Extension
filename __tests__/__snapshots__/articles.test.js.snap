// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`articles fetchArticles should fetch and store all articles on first run 1`] = `
[
  [
    {
      "articles": [
        {
          "date": "2024-06-03T22:03:20",
          "id": 283465,
          "link": "https://www.omgubuntu.co.uk/2024/06/linux-mint-will-hide-unverified-flatpaks",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/02/linux-mint-22-tile-350x200.jpg",
          "title": "Linux Mint Will Hide Unverified Flatpaks in Software Manager",
          "unread": true,
        },
        {
          "date": "2024-06-02T18:00:00",
          "id": 283423,
          "link": "https://www.omgubuntu.co.uk/2024/06/convert-text-documents-easily-ubuntu",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/06/document-conversion-350x200.jpg",
          "title": "New App Makes Converting Text Documents on Ubuntu Easier",
          "unread": true,
        },
        {
          "date": "2024-05-30T21:22:00",
          "id": 283383,
          "link": "https://www.omgubuntu.co.uk/2024/05/wsl-gui-settings-apps-coming-soon",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/windows-tux-thumbnail-350x200.jpg",
          "title": "Microsoft Announce WSL Updates, Including New Settings App",
          "unread": true,
        },
        {
          "date": "2024-05-29T22:20:00",
          "id": 283328,
          "link": "https://www.omgubuntu.co.uk/2024/05/8bitdo-keyboard-m-ibm-model-m-inspired",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/8bitdo-keyboard-m-350x200.jpg",
          "title": "IBM&#8217;s Iconic Model-M Keyboard Reimagined by 8BitDo",
          "unread": true,
        },
        {
          "date": "2024-05-28T21:48:00",
          "id": 283303,
          "link": "https://www.omgubuntu.co.uk/2024/05/firefox-126-bug-fix-drag-and-drop-linux",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2022/06/firefox-logo-on-an-orange-background-350x200.jpg",
          "title": "Mozilla Firefox 126.0.1 Fixes Drag and Drop Quirk on Linux",
          "unread": true,
        },
        {
          "date": "2024-05-28T19:35:16",
          "id": 283253,
          "link": "https://www.omgubuntu.co.uk/2024/05/ubuntu-24-04-risc-v-computer-milkv",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/milkv-mars-ubuntu-24.04-350x200.jpg",
          "title": "Ubuntu 24.04 Arrives on Mars – the Milk-V Mars (RISC-V Computer)",
          "unread": true,
        },
        {
          "date": "2024-05-27T22:23:00",
          "id": 283168,
          "link": "https://www.omgubuntu.co.uk/2024/05/day-progress-time-tracking-gnome-extension",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/day-progress-gnome-shell-extnsion-350x200.jpg",
          "title": "Track Time Differently with &#8216;Day Progress&#8217; for GNOME Shell",
          "unread": true,
        },
        {
          "date": "2024-05-26T17:21:11",
          "id": 283136,
          "link": "https://www.omgubuntu.co.uk/2024/05/cascadia-code-font-update-large-type-pieces",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/cascadia-code-update-350x200.jpg",
          "title": "Microsoft Gives its Open-Source Developer Font Major Update",
          "unread": true,
        },
        {
          "date": "2024-05-24T14:07:37",
          "id": 283080,
          "link": "https://www.omgubuntu.co.uk/2024/05/rufus-usb-ubuntu-24-04-persistence",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2016/04/usb-live-boot-350x200.jpg",
          "title": "Rufus Fixes Creation of Persistent Ubuntu 24.04 USBs",
          "unread": true,
        },
        {
          "date": "2024-05-23T15:02:35",
          "id": 283052,
          "link": "https://www.omgubuntu.co.uk/2024/05/snap-store-website-new-look",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2023/05/ubuntu-logo-and-the-snapcraft-logo-tiled-350x200.jpg",
          "title": "Snap Store Website Redesign Goes Live – And It Looks Good",
          "unread": true,
        },
        {
          "date": "2024-05-22T16:51:26",
          "id": 282972,
          "link": "https://www.omgubuntu.co.uk/2024/05/ubuntu-24-04-upgrades-officially-open",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/noble-upgrade-tile-350x200.jpg",
          "title": "Ubuntu 24.04 LTS Upgrades Now Officially Enabled",
          "unread": true,
        },
        {
          "date": "2024-05-22T15:45:05",
          "id": 282962,
          "link": "https://www.omgubuntu.co.uk/2024/05/mozilla-shares-firefox-roadmap",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2019/11/firefox-1-350x200.jpg",
          "title": "Mozilla Reveals New Features Coming to Firefox",
          "unread": true,
        },
      ],
    },
  ],
]
`;

exports[`articles fetchArticles should handle missing thumbnails 1`] = `
[
  [
    {
      "articles": [
        {
          "date": "2024-06-03T22:03:20",
          "id": 283465,
          "link": "https://www.omgubuntu.co.uk/2024/06/linux-mint-will-hide-unverified-flatpaks",
          "thumbnail": null,
          "title": "Linux Mint Will Hide Unverified Flatpaks in Software Manager",
          "unread": true,
        },
      ],
    },
  ],
]
`;

exports[`articles fetchArticles should merge new articles with existing articles 1`] = `
[
  [
    {
      "articles": [
        {
          "date": "2024-06-03T22:03:20",
          "id": 283465,
          "link": "https://www.omgubuntu.co.uk/2024/06/linux-mint-will-hide-unverified-flatpaks",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/02/linux-mint-22-tile-350x200.jpg",
          "title": "Linux Mint Will Hide Unverified Flatpaks in Software Manager",
          "unread": true,
        },
        {
          "date": "2024-06-02T18:00:00",
          "id": 283423,
          "link": "https://www.omgubuntu.co.uk/2024/06/convert-text-documents-easily-ubuntu",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/06/document-conversion-350x200.jpg",
          "title": "New App Makes Converting Text Documents on Ubuntu Easier",
          "unread": false,
        },
        {
          "date": "2024-05-30T21:22:00",
          "id": 283383,
          "link": "https://www.omgubuntu.co.uk/2024/05/wsl-gui-settings-apps-coming-soon",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/windows-tux-thumbnail-350x200.jpg",
          "title": "Microsoft Announce WSL Updates, Including New Settings App",
          "unread": false,
        },
        {
          "date": "2024-05-29T22:20:00",
          "id": 283328,
          "link": "https://www.omgubuntu.co.uk/2024/05/8bitdo-keyboard-m-ibm-model-m-inspired",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/8bitdo-keyboard-m-350x200.jpg",
          "title": "IBM&#8217;s Iconic Model-M Keyboard Reimagined by 8BitDo",
          "unread": false,
        },
        {
          "date": "2024-05-28T21:48:00",
          "id": 283303,
          "link": "https://www.omgubuntu.co.uk/2024/05/firefox-126-bug-fix-drag-and-drop-linux",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2022/06/firefox-logo-on-an-orange-background-350x200.jpg",
          "title": "Mozilla Firefox 126.0.1 Fixes Drag and Drop Quirk on Linux",
          "unread": false,
        },
        {
          "date": "2024-05-28T19:35:16",
          "id": 283253,
          "link": "https://www.omgubuntu.co.uk/2024/05/ubuntu-24-04-risc-v-computer-milkv",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/milkv-mars-ubuntu-24.04-350x200.jpg",
          "title": "Ubuntu 24.04 Arrives on Mars – the Milk-V Mars (RISC-V Computer)",
          "unread": false,
        },
        {
          "date": "2024-05-27T22:23:00",
          "id": 283168,
          "link": "https://www.omgubuntu.co.uk/2024/05/day-progress-time-tracking-gnome-extension",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/day-progress-gnome-shell-extnsion-350x200.jpg",
          "title": "Track Time Differently with &#8216;Day Progress&#8217; for GNOME Shell",
          "unread": false,
        },
        {
          "date": "2024-05-26T17:21:11",
          "id": 283136,
          "link": "https://www.omgubuntu.co.uk/2024/05/cascadia-code-font-update-large-type-pieces",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/cascadia-code-update-350x200.jpg",
          "title": "Microsoft Gives its Open-Source Developer Font Major Update",
          "unread": false,
        },
        {
          "date": "2024-05-24T14:07:37",
          "id": 283080,
          "link": "https://www.omgubuntu.co.uk/2024/05/rufus-usb-ubuntu-24-04-persistence",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2016/04/usb-live-boot-350x200.jpg",
          "title": "Rufus Fixes Creation of Persistent Ubuntu 24.04 USBs",
          "unread": false,
        },
        {
          "date": "2024-05-23T15:02:35",
          "id": 283052,
          "link": "https://www.omgubuntu.co.uk/2024/05/snap-store-website-new-look",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2023/05/ubuntu-logo-and-the-snapcraft-logo-tiled-350x200.jpg",
          "title": "Snap Store Website Redesign Goes Live – And It Looks Good",
          "unread": false,
        },
        {
          "date": "2024-05-22T16:51:26",
          "id": 282972,
          "link": "https://www.omgubuntu.co.uk/2024/05/ubuntu-24-04-upgrades-officially-open",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2024/05/noble-upgrade-tile-350x200.jpg",
          "title": "Ubuntu 24.04 LTS Upgrades Now Officially Enabled",
          "unread": false,
        },
        {
          "date": "2024-05-22T15:45:05",
          "id": 282962,
          "link": "https://www.omgubuntu.co.uk/2024/05/mozilla-shares-firefox-roadmap",
          "thumbnail": "https://www.omgubuntu.co.uk/wp-content/uploads/2019/11/firefox-1-350x200.jpg",
          "title": "Mozilla Reveals New Features Coming to Firefox",
          "unread": false,
        },
      ],
    },
  ],
]
`;
